docs/specification/revisions/draft/auth/oauth2.md

```diff
@@ -0,0 +1,243 @@
+---
+title: OAuth 2.0 Authentication
+type: docs
+weight: 51
+---
+
+{{< callout type="warning" >}}
+Auth is **experimental**, and being drafted for release in the next [revision]({{< ref "/specification/revisions" >}}) of the protocol.
+
+The additions to the base protocol are backwards compatible to revision 2024-11-05; however, **the auth specification may change in backwards incompatible ways** until the next protocol revision.
+{{< /callout >}}
+
+The Model Context Protocol (MCP) supports [OAuth 2.0](https://oauth.net/2/) as a standardized authentication method, allowing secure authorization flows between clients and servers. Tokens will be securely communicated as part of the request body.
+
+## Protocol Flow
+We follow the flows defined by [RFC 6749](https://datatracker.ietf.org/doc/html/rfc6749).
+
+```mermaid
+sequenceDiagram
+    participant Client as MCP Client
+    participant User as Resource Owner (End User)
+    participant AuthServer as Authorization<br> MCP Server
+    participant Resource as Resource<br>MCP Server
+
+    Note right of Client: Obtaining an access token:
+    Client->>User: Authorization Request (auth/oauth2/authorize)
+    User->>Client: Authorization Grant
+    Client->>AuthServer: Authorization Grant (auth/oauth2/token)
+    AuthServer->>Client: Access Token
+
+    Note right of Client: Using an access token:
+    Client->>Resource: Access prompts/resources/tools with Access Token
+    Resource->>Client: Receive protected response
+    
+
+    Note right of Client: Refreshing an access token:
+    Client->>AuthServer: Refresh Token Request (auth/oauth2/token)
+    AuthServer->>Client: New Access Token
+
+    Note right of Client: Revoking tokens:
+    Client->>AuthServer: Revoke Token Request (auth/oauth2/revoke)
+```
+Note that the authorization server may be the same server as the resource server or a separate entity. A single authorization server may issue access tokens accepted by multiple resource servers.
+
+## Capabilities
+
+
+Servers supporting OAuth 2.0 **MUST** include their capabilities:
+
+```json
+{
+  "capabilities": {
+    "auth": {
+      "oauth2": {
+        "authorize": {},
+        "token": {},
+        "revoke": {}
+      }
+    }
+  }
+}
+```
+
+## Flows
+### Initialization
+During initialization, if the server supports the `oauth2` capability, the client **SHOULD** include an access token in all subsequent requests. If the client does not have an access token, the client **SHOULD** obtain one.
+
+### Obtaining an Access Token
+
+#### Authorization Grant Flow
+To obtain an access token, clients **MUST** send an `auth/oauth2/authorize` request.
+
+**Request:**
+```typescript
+{
+  "jsonrpc": "2.0",
+  "id": 1,
+  "method": "auth/oauth2/authorize",
+  "params": {
+    "response_type": "code", // REQUIRED
+    "client_id": "client_id", // REQUIRED
```

Comment:
How does the MCP Client obtain the `client_id` value? And, what type of OAuth2 client is the MCP Client?

It seems like there are two possible approaches:

1. Treat the MCP Client as a [public client](https://datatracker.ietf.org/doc/html/rfc6749#section-2.1) and use the Authorization Code Flow with PKCE. However, you still need to clarify how the `client_id` is assigned or obtained in this case.
2. Treat the MCP Client as a third-party client, where clients are created through one of the following methods:
2.1. Using the [OAuth 2.0 Dynamic Client Registration Protocol](https://datatracker.ietf.org/doc/html/rfc7591) (disadvantage: not many Authorization Servers seem to support this).
2.2. Leveraging a developer portal provided by the Authorization Server.